File: vlc-android/src/org/videolan/vlc/gui/browser/StorageBrowserAdapter.java
/*
 * *************************************************************************
 *  StorageBrowserAdapter.java
 * **************************************************************************
 *  Copyright © 2015 VLC authors and VideoLAN
 *  Author: Geoffrey Métais
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston MA 02110-1301, USA.
 *  ***************************************************************************
 */

package org.videolan.vlc.gui.browser;

import android.content.SharedPreferences;
import android.net.Uri;
import android.preference.PreferenceManager;
import android.view.View;
import android.widget.CheckBox;

import org.videolan.medialibrary.media.MediaLibraryItem;
import org.videolan.medialibrary.media.MediaWrapper;
import org.videolan.medialibrary.media.Storage;
import org.videolan.vlc.VLCApplication;
import org.videolan.vlc.gui.helpers.MedialibraryUtils;
import org.videolan.vlc.gui.helpers.ThreeStatesCheckbox;
import org.videolan.vlc.util.Constants;
import org.videolan.vlc.util.CustomDirectories;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

class StorageBrowserAdapter extends BaseBrowserAdapter {

    private static List<String> mMediaDirsLocation;
    private static List<String> mCustomDirsLocation;

    StorageBrowserAdapter(BaseBrowserFragment fragment) {
        super(fragment);
        updateMediaDirs();
    }

    @Override
    public void onBindViewHolder(final ViewHolder holder, int position) {
        final MediaViewHolder vh = (MediaViewHolder) holder;
        MediaLibraryItem storage = getItem(position);

        if (storage.getItemType() == MediaLibraryItem.TYPE_MEDIA) storage = new Storage(((MediaWrapper)storage).getUri());
        String storagePath = ((Storage)storage).getUri().getPath();
        if (!storagePath.endsWith("/")) storagePath += "/";
        boolean hasContextMenu = mCustomDirsLocation.contains(storagePath);
        boolean checked = ((StorageBrowserFragment) fragment).mScannedDirectory || mMediaDirsLocation.contains(storagePath);
        vh.binding.setItem(storage);
        vh.binding.setHasContextMenu(hasContextMenu);
        if (checked)
            vh.binding.browserCheckbox.setState(ThreeStatesCheckbox.STATE_CHECKED);
        else if (hasDiscoveredChildren(storagePath))
            vh.binding.browserCheckbox.setState(ThreeStatesCheckbox.STATE_PARTIAL);
        else
            vh.binding.browserCheckbox.setState(ThreeStatesCheckbox.STATE_UNCHECKED);
        vh.binding.setCheckEnabled(!((StorageBrowserFragment) fragment).mScannedDirectory);
    }

    private boolean hasDiscoveredChildren(String path) {
        for (String directory : mMediaDirsLocation) if (directory.startsWith(path)) return true;
        return false;
    }

    void updateMediaDirs() {
        if (mMediaDirsLocation != null) mMediaDirsLocation.clear();
        final String folders[] = VLCApplication.getMLInstance().getFoldersList();
        mMediaDirsLocation = new ArrayList<>(folders.length);
        for (String folder : folders) {
            mMediaDirsLocation.add(Uri.decode(folder.startsWith("file://") ? folder.substring(7) : folder));
        }
        mCustomDirsLocation = new ArrayList<>(Arrays.asList(CustomDirectories.getCustomDirectories()));
    }

    protected void checkBoxAction(View v, String mrl) {
        final ThreeStatesCheckbox tscb = (ThreeStatesCheckbox) v;
        int state = tscb.getState();
        if (state == ThreeStatesCheckbox.STATE_CHECKED) {
            MedialibraryUtils.addDir(mrl);
            final SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(v.getContext().getApplicationContext());
            if (prefs.getInt(Constants.KEY_MEDIALIBRARY_SCAN, -1) != Constants.ML_SCAN_ON) prefs.edit().putInt(Constants.KEY_MEDIALIBRARY_SCAN, Constants.ML_SCAN_ON).apply();
        } else MedialibraryUtils.removeDir(mrl);
        ((StorageBrowserFragment)fragment).processEvent((CheckBox) v, mrl);
    }
}

File: vlc-android/src/org/videolan/vlc/gui/browser/StorageBrowserAdapter.kt
/*
 * *************************************************************************
 *  StorageBrowserAdapter.java
 * **************************************************************************
 *  Copyright © 2015 VLC authors and VideoLAN
 *  Author: Geoffrey Métais
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston MA 02110-1301, USA.
 *  ***************************************************************************
 */

package org.videolan.vlc.gui.browser

import android.content.Context
import android.net.Uri
import android.view.View
import androidx.databinding.ViewDataBinding
import kotlinx.coroutines.*
import org.videolan.medialibrary.interfaces.AbstractMedialibrary
import org.videolan.medialibrary.interfaces.media.AbstractMediaWrapper
import org.videolan.medialibrary.media.MediaLibraryItem
import org.videolan.medialibrary.media.Storage
import org.videolan.vlc.MediaParsingService
import org.videolan.vlc.gui.helpers.ThreeStatesCheckbox
import org.videolan.vlc.repository.DirectoryRepository
import org.videolan.vlc.util.containsPath

@ObsoleteCoroutinesApi
@ExperimentalCoroutinesApi
internal class StorageBrowserAdapter(fragment: StorageBrowserFragment) : BaseBrowserAdapter(fragment) {

    private var mediaDirsLocation: MutableList<String> = mutableListOf()
    private lateinit var customDirsLocation: List<String>
    private var updateJob : Job? = null

    init {
        updateMediaDirs(fragment.requireContext())
    }

    override fun onBindViewHolder(holder: ViewHolder<ViewDataBinding>, position: Int) {
        val vh = holder as MediaViewHolder
        vh.job = launch {
            var storage = getItem(position)
            if (storage.itemType == MediaLibraryItem.TYPE_MEDIA) storage = Storage((storage as AbstractMediaWrapper).uri)
            var storagePath = (storage as Storage).uri.path ?: ""
            if (!storagePath.endsWith("/")) storagePath += "/"
            vh.binding.item = storage
            updateJob?.join()
            if (updateJob?.isCancelled == true) return@launch
            val hasContextMenu = customDirsLocation.contains(storagePath)
            val checked = (fragment as StorageBrowserFragment).scannedDirectory || mediaDirsLocation.containsPath(storagePath)
            vh.binding.hasContextMenu = hasContextMenu
            when {
                checked -> vh.binding.browserCheckbox.state = ThreeStatesCheckbox.STATE_CHECKED
                hasDiscoveredChildren(storagePath) -> vh.binding.browserCheckbox.state = ThreeStatesCheckbox.STATE_PARTIAL
                else -> vh.binding.browserCheckbox.state = ThreeStatesCheckbox.STATE_UNCHECKED
            }
            vh.binding.checkEnabled = !fragment.scannedDirectory
        }
    }

    override fun onViewRecycled(holder: ViewHolder<ViewDataBinding>) {
        (holder as MediaViewHolder).apply {
            job?.cancel()
            job = null
        }
        super.onViewRecycled(holder)
    }

    private fun hasDiscoveredChildren(path: String): Boolean {
        for (directory in mediaDirsLocation) if (directory.startsWith(path)) return true
        return false
    }

    suspend fun updateListState(context: Context) {
        updateMediaDirs(context)
        updateJob?.join()
        if (updateJob?.isCancelled == false) notifyItemRangeChanged(0, itemCount)
    }

    fun updateMediaDirs(context: Context) {
        mediaDirsLocation.clear()

        updateJob = launch {
            val folders = if (!AbstractMedialibrary.getInstance().isInitiated) {
                MediaParsingService.preselectedStorages.toTypedArray()
            } else {
                withContext(Dispatchers.IO) { AbstractMedialibrary.getInstance().foldersList }
            }

            folders.forEach {
                mediaDirsLocation.add(Uri.decode(if (it.startsWith("file://")) it.substring(7) else it))
            }
            customDirsLocation = DirectoryRepository.getInstance(context).getCustomDirectories().map { it.path }
        }
    }

    override fun checkBoxAction(v: View, mrl: String) {
        (fragment as? StorageBrowserFragment)?.checkBoxAction(v, mrl)
    }
}
